capsules/src/net/lowpan_fragment.rs

```diff
@@ -0,0 +1,900 @@
+        self.rx_states.push_head(rx_state);
+    }
+
+    /// FragState::set_receive_client
+    /// -----------------------------
+    /// This function sets the client to receive the callback when a packet
+    /// has been fully reassembled. In the current design, the concept of a
+    /// receive client per RxState did not make sense; thus, there is a single
+    /// receive client that receives all reassembled packets.
+    pub fn set_receive_client(&self, client: &'static ReceiveClient) {
+        self.rx_client.set(Some(client));
+    }
+
+    /// FragState::transmit_packet
+    /// --------------------------
+    /// This function is called to send a fully-formed IPv6 packet. Arguments
+    /// to this function are used to determine various aspects of the MAC
+    /// layer frame and keep track of the transmission state.
+    pub fn transmit_packet(&self,
+                           src_mac_addr: MacAddress,
+                           dst_mac_addr: MacAddress,
+                           mut ip6_packet: &'static mut [u8],
+                           ip6_packet_len: usize,
+                           security: Option<(SecurityLevel, KeyId)>,
+                           tx_state: &'a TxState<'a>,
+                           fragment: bool,
+                           compress: bool)
+                           -> Result<ReturnCode, ReturnCode> {
+
+        tx_state.init_transmit(src_mac_addr,
+                               dst_mac_addr,
+                               ip6_packet,
+                               ip6_packet_len,
+                               security,
+                               fragment,
+                               compress);
+        // Queue tx_state
+        self.tx_states.push_tail(tx_state);
+        if self.tx_busy.get() {
+            Ok(ReturnCode::SUCCESS)
+        } else {
+            // Set as current state and start transmit
+            self.start_packet_transmit();
+            Ok(ReturnCode::SUCCESS)
+        }
+    }
+
+    fn start_packet_transmit(&self) {
+        // Already transmitting
+        if self.tx_busy.get() {
+            return;
+        }
+
+        let dgram_tag = if (self.tx_dgram_tag.get() + 1) == 0 {
+            1
+        } else {
+            self.tx_dgram_tag.get() + 1
+        };
+
+        while self.tx_states.head().map_or(false, move |state| {
+            // We panic here, as it should never be the case that we start
+            // transmitting without the tx_buf, since the `tx_buf` is owned by
+            // the TxState struct and is passed to the radio during transmission.
+            // Failure to replace the `tx_buf` or failure to initialize TxState
+            // with a `tx_buf` represents a significant logic error, and we should
+            // panic.
+            let mut frag_buf = self.tx_buf
+                .take()
+                .expect("Error: `tx_buf` is None in call to start_packet_transmit.");
```

Comment:
The expect call inside `start_packet_transmit` (line 734) is again indicative of a logic bug. Since the `tx_buf` is owned by the `FragState` struct and passed between the radio and the fragmentation layer, if we are not busy and the `tx_buf` buffer is `None`, then we have lost the buffer, and can therefore never transmit.